services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala

```diff
@@ -0,0 +1,59 @@
+package cromwell.services.auth.ecm
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.RawHeader
+import akka.util.ByteString
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Success, Try}
+import spray.json._
+
+class EcmService(baseEcmUrl: String) {
+  private val getGithubAccessTokenApiPath = "api/oauth/v1/github/access-token"
+
+  /*
+     ECM doesn't have a standard error response format. Some of the responses contains HTML tags in it. This helper
+     method returns custom error message for 401 and 403 errors as they contain HTML tags. For 400, 404 and 500 the
+     Swagger suggests that the response format is of ErrorReport schema and this method tries to extract the
+     actual message from the JSON object and returns it. In case of other status codes or if it fails to parse JSON it
+     returns the original error response.
+     ErrorReport schema: {"message":<actual_error_msg>, "statusCode":<code>}
+   */
+  def extractErrorMessage(errorCode: StatusCode, responseBodyAsStr: String): String =
+    errorCode match {
+      case StatusCodes.Unauthorized => "Invalid or missing authentication credentials."
+      case StatusCodes.Forbidden => "User doesn't have the right permission(s) to fetch Github token."
+      case StatusCodes.BadRequest | StatusCodes.NotFound | StatusCodes.InternalServerError =>
+        Try(responseBodyAsStr.parseJson) match {
+          case Success(JsObject(fields)) =>
+            fields.get("message").map(_.toString().replaceAll("\"", "")).getOrElse(responseBodyAsStr)
+          case _ => responseBodyAsStr
+        }
+      case _ => responseBodyAsStr
+    }
+
+  def getGithubAccessToken(
+    userToken: String
+  )(implicit actorSystem: ActorSystem, ec: ExecutionContext): Future[String] = {
+
+    def responseEntityToFutureStr(responseEntity: ResponseEntity): Future[String] =
+      responseEntity.dataBytes.runFold(ByteString(""))(_ ++ _).map(_.utf8String)
```

Comment:
I know akka is picky and can throw exceptions if we don't read the bytes in time (within 1 second?). How comfortable are we that this future will evaluate in time?

Reply:
I copied the pattern we have in Cromwell (see [WorkflowCallbackActor.scala](https://github.com/broadinstitute/cromwell/blob/027de62cf8e37f62330d76dba67ed07cce81a1ba/engine/src/main/scala/cromwell/engine/workflow/lifecycle/finalization/WorkflowCallbackActor.scala#L210) and [TesAsyncBackendJobExecutionActor.scala](https://github.com/broadinstitute/cromwell/blob/027de62cf8e37f62330d76dba67ed07cce81a1ba/supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala#L490)) and didn't realize that 1 second was the default timeout. Is there an implicit timeout being imported in these 2 references that I also need to update here? 🤔

Reply:
It's possible that we're hitting this error in cases where we aren't choosing to read the bytes at all, rather than cases where we take too long to read them. We're planning to do [WX-1525](https://broadworkbench.atlassian.net/browse/WX-1525) next week, which should confirm that one way or the other.

Reply:
In this case, should the whole response be loaded into memory as `Strict` entity using `.toStrict(<timeout>)` method?

Reply:
I would just make sure that `responseEntity.dataBytes.runFold` is actually getting executed in all code paths. It looks like you're calling it on both success and failure codes, so I think you're probably okay!